components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,260 @@
+                &[&*observer_size],
+            );
+            entries.push(entry);
+            observation
+                .last_reported_sizes
+                .borrow_mut()
+                .push_back(size_impl);
+            observation.state = ObservationState::Done;
+            let target_depth = calculate_depth_for_node(target);
+            if target_depth < *shallowest_target_depth {
+                *shallowest_target_depth = target_depth;
+            }
+        }
+        let _ = self
+            .callback
+            .Call__(entries, self, ExceptionHandling::Report);
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#has-skipped-observations-h>
+    pub fn has_skipped_resize_observations(&self) -> bool {
+        self.observation_targets
+            .borrow()
+            .iter()
+            .any(|(obs, _)| matches!(obs.state, ObservationState::Skipped))
+    }
+}
+
+impl ResizeObserverMethods for ResizeObserver {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-observe>
+    fn Observe(&self, target: &Element, options: &ResizeObserverOptions) {
+        let is_present = self
+            .observation_targets
+            .borrow()
+            .iter()
+            .any(|(_obs, other)| &**other == target);
+        if is_present {
+            self.Unobserve(target);
+        }
+
+        let resize_observation = ResizeObservation::new(options.box_);
+
+        self.observation_targets
+            .borrow_mut()
+            .push((resize_observation, Dom::from_ref(target)));
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-unobserve>
+    fn Unobserve(&self, target: &Element) {
+        self.observation_targets
+            .borrow_mut()
+            .retain_mut(|(_obs, other)| !(&**other == target));
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-disconnect>
+    fn Disconnect(&self) {
+        self.observation_targets.borrow_mut().clear();
+    }
+}
+
+/// State machine equivalent of active and skipped observations.
+#[derive(Default, JSTraceable, MallocSizeOf)]
+enum ObservationState {
+    #[default]
+    Done,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-activetargets-slot>
+    Active,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-skippedtargets-slot>
+    Skipped,
+}
+
+/// https://drafts.csswg.org/resize-observer/#resizeobservation
+#[derive(JSTraceable, MallocSizeOf)]
+struct ResizeObservation {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-target>
+    /// Note: `target` is kept out of here, to avoid having to root the `ResizeObservation`.
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-observedbox>
+    observed_box: ResizeObserverBoxOptions,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-lastreportedsizes>
+    last_reported_sizes: DomRefCell<VecDeque<ResizeObserverSizeImpl>>,
```

Comment:
Yeah nothing... I was thinking of making this just an option, because only the last one is ever read. What do you think?

Reply:
An option is fine as long as we only report one size at a time. you really want to set this to the array we last reported, rather than keeping appending.

Reply:
Ah ok, now I understand: it's the array for multiple fragments for a single reported size, where currently one fragment is used. I thought it was more of a stack of reported sizes... thanks!